build-system/tasks/e2e/README.md

```diff
@@ -0,0 +1,181 @@
+  expect(text).to.equal('baz');
+});
+```
+
+### implicit polling
+
+With an asynchronous `expect`, the test can wait for the actual to match the expected value. For example, the code above will still work even if the test begins before the component is able to update the `fooBars[0]` element's text to `'baz'`.
+
+To enable implicit polling, the AMP E2E framework implements `ControllerPromises` that allow code that is aware of them to request an updated value from the test controller. `expect` will implicitly wait if it is passed a `ControllerPromise` directly. If a `ControllerPromise` is `await`ed, the value will be retrieved immediately without polling, and so they should not be `await`ed if polling is desired.
+
+```js
+it('should render correctly', async() => {
+  const fooBars = await controller.findElements('.i-amphtml-foo-bar');
+
+  // OK: waits for the text to equal 'baz'
+  await expect(controller.getElementText(fooBars[0])).to.equal('baz');
+
+  // OK: waits for the text to equal 'baz'
+  const text = controller.getElementText(fooBars[0]);
+  await expect(text).to.equal('baz');
+
+  // Avoid: expects the value to immediately equal 'baz'
+  await expect(await controller.getElementText(fooBars[0])).to.equal('baz');
+
+  // Avoid: expects the value to immediately equal 'baz'
+  const text = await controller.getElementText(fooBars[0]);
+  await expect(text).to.equal('baz');
+});
+```
+
+
+## Debugging E2E tests
+
+E2E test code runs in `node` and the code under test runs in the browser with the Web Driver in between. This indirection adds a level of difficulty to debugging, but the following commands work well to resolve any confusion or issues that arise while writing and debugging tests.
+
+### node inspect
+
+```sh
+node --inspect-brk $(which gulp) e2e # --nobuild --testnames --files=extensions/amp-foo/0.1/test-e2e/test-amp-foo-basic.js
+```
+
+Open Chrome DevTools and click the Node logo in the top left.
+
+Click the "Play"/"Continue execution" button to continue execution. Execution will stop at `debugger` statements like they would in the browser.
+
+### mocha only
+
+`it.only` and `describe.only` make it easy to execute individual tests or groups of tests.
+
+### E2E repl (experimental)
+
+Add `repl(/* mochaThis */ this)` to a test to pause execution at that line without using a breakpoint or `debugger` statement.
+
+`debugger` statements can be troublesome because they stop all code execution. When the debugger is paused, synchronous calls in the console return values, but asynchronous calls return `Promise (unresolved)` which makes debugging expected values difficult at breakpoints. `repl` stops execution from continuing, but allows Promises to resolve normally.
+
+```diff
+-  it('should render correctly', async() => {
++  it('should render correctly', async function() {
+     const fooBars = await controller.findElements('.i-amphtml-foo-bar');
+     await expect(fooBars).to.have.length(2);
++    await repl(this);
```

Comment:
how do I continue?

Reply:
good question, I'll add a reference to `repl.continue()` in the explanation